src/Type/Php/ArrayKeysFunctionDynamicReturnTypeExtension.php

```diff
@@ -36,7 +40,22 @@ public function getTypeFromFunctionCall(FunctionReflection $functionReflection,
 			return $this->phpVersion->arrayFunctionsReturnNullWithNonArray() ? new NullType() : new NeverType();
 		}
 
-		return $arrayType->getKeysArray();
+		$keysArray = $arrayType->getKeysArray();
+		if (count($functionCall->getArgs()) === 1) {
+			return $keysArray;
+		}
+
+		$newArrayType = $keysArray;
+		if (!$keysArray->isConstantArray()->no()) {
+			$newArrayType = new ArrayType(
+				$keysArray->getIterableKeyType()->generalize(GeneralizePrecision::lessSpecific()),
+				$keysArray->getIterableValueType()->generalize(GeneralizePrecision::lessSpecific()),
+			);
+		}
```

Comment:
we could make this more sophisticated and support more filtering on constant array, but I think this is good enough for now.

until a few days ago I didn't even know `array_keys` can filter by values - it feels pretty rare